File: backend-crud-CSharp/MessageQueueService/MessageQueueService.cs
using System;
using System.Text;
using System.Collections.Concurrent;

using RabbitMQ.Client;
using RabbitMQ.Client.Events;

namespace RebelSoftware.MessageQueue
{
    class RPCInfo
    {
        public IConnection Connection { get; set; }
        public IModel Channel { get; set; }
        public EventingBasicConsumer Consumer { get; set; }
        public BlockingCollection<string> RespQueue { get; set; }
        public IBasicProperties Props { get; set; }
    }

    internal class MessageQueueService : IMessageQueueService
    {
        private void onMessageReceived (BlockingCollection<string> respQueue, string correlationId, BasicDeliverEventArgs ea)
        { 
            var response = Encoding.UTF8.GetString(ea.Body);
            if(ea.BasicProperties.CorrelationId == correlationId)
            {
                respQueue.Add(response);
            }
        }

        private RPCInfo CreateRpcInfoObject()
        { 
            var factory = new ConnectionFactory();
            factory.HostName = "localhost"; //todo: caller gets host name from config file and passes it to createMessageQueuer
            var respQueue = new BlockingCollection<string>();
            var connection = factory.CreateConnection();
            var channel = connection.CreateModel();
            var consumer = new EventingBasicConsumer(channel);
            var props = channel.CreateBasicProperties();
            props.CorrelationId = Guid.NewGuid().ToString();
            props.ReplyTo = channel.QueueDeclare().QueueName;
            //consumer.Received.Add(onMessageReceived, respQueue, props.CorrelationId);
            consumer.Received += (sender, eventArgs) => {
                onMessageReceived(respQueue, props.CorrelationId, eventArgs);
            };
            
            var retVal = new RPCInfo
            {
                Connection = connection,
                Channel = channel,
                Consumer = consumer,
                RespQueue = respQueue,
                Props = props
            };

            return retVal;
        }

        private void PublishToMsgQueue (string jsonString, RPCInfo rpcInfo)
        { 
            var msgBytes = Encoding.UTF8.GetBytes(jsonString);
            rpcInfo.Channel.BasicPublish(exchange: String.Empty, routingKey: "employee", basicProperties: rpcInfo.Props, body: msgBytes); //todo: routing key from config file
            rpcInfo.Channel.BasicConsume(consumer: rpcInfo.Consumer, queue: rpcInfo.Props.ReplyTo, autoAck: true);
        }

        public string WriteMessageAndGetResponse(string message)
        {
            var rpcInfo = CreateRpcInfoObject();
            PublishToMsgQueue(message, rpcInfo);

            var mqResponse = String.Empty;
            if(rpcInfo.RespQueue.TryTake(out mqResponse, 30000))
            {
                return mqResponse;
            } //todo: timeout value from config file
            else
            {
                var errMsg = $"timeout while waiting for microservice response to RabbitMQ message '{message}'";
                throw new Exception(errMsg);
            }
        }
    }
}

File: backend-crud-Functional/MessageQueueService/MessageQueueService.fs
namespace RebelSoftware.MessageQueueService //todo: rename to RebelSoftware.MessageQueue

open System
open System.Text
open System.Collections.Concurrent

open RabbitMQ.Client
open RabbitMQ.Client.Events

module MessageQueueing =
    type private RPCInfo = {
        //todo: capitalize member names
        connection : IConnection;
        channel : IModel;
        consumer : EventingBasicConsumer;
        respQueue : BlockingCollection<string>;
        props : IBasicProperties;
    }

    let private onMessageReceived (respQueue:BlockingCollection<string>) correlationId (ea:BasicDeliverEventArgs) = 
        let response = Encoding.UTF8.GetString(ea.Body)
        if(ea.BasicProperties.CorrelationId = correlationId) then
            respQueue.Add(response)

    let private createRpcInfoObject () = 
        let factory = ConnectionFactory()
        factory.HostName <- "localhost" //todo: does F# have object initializer? //todo: caller gets host name from config file and passes it to createMessageQueuer
        let respQueue = new BlockingCollection<string>()
        let connection = factory.CreateConnection()
        let channel = connection.CreateModel()
        let consumer = EventingBasicConsumer(channel)
        let props = channel.CreateBasicProperties();
        props.CorrelationId <- Guid.NewGuid().ToString()
        props.ReplyTo <- channel.QueueDeclare().QueueName
        consumer.Received.Add(onMessageReceived respQueue props.CorrelationId)
        {
            connection = connection;
            channel = channel;
            consumer = consumer;
            respQueue = respQueue;
            props = props;
        }
    
    let private publishToMsgQueue (jsonString:string) rpcInfo = 
        let msgBytes = Encoding.UTF8.GetBytes(jsonString)
        rpcInfo.channel.BasicPublish(exchange=String.Empty, routingKey="employee", basicProperties=rpcInfo.props, body=msgBytes) //todo: routing key from config file
        rpcInfo.channel.BasicConsume(consumer = rpcInfo.consumer, queue=rpcInfo.props.ReplyTo, autoAck=true)

    let writeMessageAndGetResponse jsonString =
        
        let rpcInfo = createRpcInfoObject () //todo 1: create MQ Layer, startup injects function pointer
        rpcInfo
        |> publishToMsgQueue jsonString 
        |> ignore

        let mutable mqResponse = String.Empty
        match rpcInfo.respQueue.TryTake(&mqResponse, 30000) with //todo: timeout value from config file
        | true -> mqResponse
        | false -> 
            sprintf "timeout while waiting for microservice response to RabbitMQ message '%s'" jsonString
            |> Exception 
            |> raise